### PR TITLE
=typ graceful stop examples (#22776)

### PR DESCRIPTION
**Pull request by gosubpl**
* Stop all the children
* Perform the cleanup
* Stop the parent

### CODE REVIEW

**Comment by gosubpl** [comment]
Refs #22776

**Comment by gosubpl** [comment]
A couple of questions, still, hence [wip]:
* Did I get the signature of `cleanup` right ?
* Shall I explicitly `ignore` messages (currently using Heiko's `signal` behaviour)
* I am "swallowing" other signals than the one I am interested in. Is this ok?
* Maybe some more tests?

**Comment by gosubpl** [comment]
Is `Jenkins` feeling fine today?
```
[error] ## Exception when compiling 194 sources to /localhome/jenkinsakka/workspace/pr-validator-per-commit-jenkins/akka-actor/target/classes
[error] java.lang.NullPointerException
[error] com.sun.tools.javac.main.Main.compile(Main.java:559)
```

**Comment by raboof** [comment]
PLS BUILD

**Comment by raboof** [comment]
@gosubpl I think that's #23859

File: akka-typed/src/main/scala/akka/typed/javadsl/Actor.scala [hunk 1]
```diff
@@ -202,6 +201,15 @@ object Actor {
   }
 
   /**
+   * Behavior that waits for all child actors to stop and then
+   * runs some cleanup before stopping.
+   */
```
Comment (ktoso) [inline akka-typed/src/main/scala/akka/typed/javadsl/Actor.scala]: Clarify that it actually stops them and awaits for them, not just awaits for it

File: akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala [hunk 2]
```diff
@@ -194,6 +193,28 @@ object Actor {
     immutable[T]((_, _) ⇒ same).onSignal(handler)
 
   /**
+   * Behavior that waits for all child actors to stop and then
+   * runs some cleanup before stopping.
```
Comment (ktoso) [inline akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala]: clarify that it performs the stopping

**Review by ktoso — CHANGES_REQUESTED** [review]
Minor comments, otherwise looks good!

**Comment by ktoso** [comment]
I wonder if it also could be mentioned in docs right away? I know we'll have to re-do docs, but this could be easy to add with the example from the doc test perhaps?

File: akka-typed/src/main/scala/akka/typed/javadsl/Actor.scala (hunk 1 above)
Comment (gosubpl) [reply akka-typed/src/main/scala/akka/typed/javadsl/Actor.scala]: done

File: akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala (hunk 2 above)
Comment (gosubpl) [reply akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala]: done

File: akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala [hunk 3]
```diff
@@ -194,6 +193,28 @@ object Actor {
...
+        case (ctx, Terminated(_)) ⇒
+          if (ctx.children.isEmpty) {
+            cleanup(ctx)
+            Behavior.stopped
+          } else
+            Behavior.same
+        case _ ⇒
+          Behavior.same
+      }
+    }
```
[the remark this answers is not in the document]
Comment (gosubpl) [reply akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala]: It is scary. And now - even works!

**Comment by gosubpl** [comment]
Review comments reflected, docs code added (both Java and Scala). Allow me to add - writing docs code has been a very good exercise, as it allowed to iron out some details.

**Comment by gosubpl** [comment]
PLS BUILD

**Comment by gosubpl** [comment]
Seems I have lost my magical powers over Jenkins :(

**Comment by gosubpl** [comment]
@ktoso comments reflected.

File: akka-typed-tests/src/test/java/jdocs/akka/typed/GracefulStopDocTest.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.concurrent.TimeUnit;
+import scala.concurrent.Await;
+import scala.concurrent.duration.Duration;
+
+public class GracefulStopDocTest {
+
+    //#master-actor
+    public abstract static class JobControl {
+        //no instances of this class, it's only a name space for messages
+        // and static methods
```
Comment (ktoso) [inline akka-typed-tests/src/test/java/jdocs/akka/typed/GracefulStopDocTest.java]: method inside? missing space before `no`?

**Review by ktoso — APPROVED** [review]
LGTM from my side; anyone else want to check @akka/akka-team

File: akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala [hunk 5]
```diff
@@ -193,6 +192,32 @@ object Actor {
   def onSignal[T](handler: PartialFunction[(ActorContext[T], Signal), Behavior[T]]): Behavior[T] =
     immutable[T]((_, _) ⇒ same).onSignal(handler)
 
+  /**
+   * Behavior that stops all child actors, waits for them to stop and then
+   * runs some user defined cleanup before stopping itself.
+   */
+  def gracefulStop[T](cleanup: ActorContext[T] ⇒ Unit): Behavior[T] =
+    Actor.deferred { context ⇒
+      context.children.foreach { cld ⇒
```
Comment (hepin1989) [inline akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala]: move these codes inside else  block

File: akka-typed-tests/src/test/java/jdocs/akka/typed/GracefulStopDocTest.java (hunk 4 above)
Comment (gosubpl) [reply akka-typed-tests/src/test/java/jdocs/akka/typed/GracefulStopDocTest.java]: fixed missing space. Didn't exactly get the point about "method inside". Would you be so kind and clarify?

File: akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala (hunk 5 above)
Comment (gosubpl) [reply akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala]: done. Thanks for spotting!

**Comment by gosubpl** [comment]
```
[info] Checking Policies for project akka-adhoc in product Lightbend Reactive Platform
[error] sbtwhitesource.WhiteSourceException: Error communicating with service: Bad Gateway
[error] 	at sbtwhitesource.CheckPoliciesAction.sendCheckPolicies(WhiteSource.scala:382)
[error] 	at sbtwhitesource.CheckPoliciesAction.doExecute(WhiteSource.scala:364)
```

**Comment by raboof** [comment]
PLS BUILD

**Comment by chbatey** [comment]
PLS BUILD

**Comment by ktoso** [comment]
What I meant was https://github.com/akka/akka/pull/23890/commits/f3d11cbc4b91eb0dd3697f88f58c62de83113808 @golem131 @gosubpl

File: akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala [hunk 6]
```diff
@@ -193,6 +192,33 @@ object Actor {
...
+        cleanup(context)
+        Behavior.stopped
+      } else {
+        context.children.foreach { cld ⇒
+          context.watch(cld)
+          context.stop(cld)
+        }
+      }
+      onSignal {
+        case (ctx, Terminated(_)) ⇒
```
Comment (johanandren) [inline akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala]: use a guard on this to get one case where we stop and one where we return `same`?

File: akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (johanandren) [reply akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala]: This is still the same, is it our auto formatting doing it, if not please format according to Konrads request.

File: akka-typed/src/main/scala/akka/typed/javadsl/Actor.scala [hunk 7]
```diff
@@ -201,6 +200,15 @@ object Actor {
     immutable(jSame, handler)
   }
 
+  /**
+   * Behavior that stops all child actors, waits for them to stop and then
+   * runs some user defined cleanup before stopping itself.
```
Comment (johanandren) [inline akka-typed/src/main/scala/akka/typed/javadsl/Actor.scala]: I'd prefer "the user defined cleanup" over "some" which sounds like random cleanup it picks up somewhere

File: akka-typed-tests/src/test/scala/docs/akka/typed/GracefulStopDocSpec.scala [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    // TODO Implicits.global is not something we would like to encourage in docs
+    //#start-workers
+    import MasterControlProgramActor._
+
+    val system: ActorSystem[JobControlLanguage] = ActorSystem(mcpa, "B6700")
+
+    system ! SpawnJob("a")
+    system ! SpawnJob("b")
+
+    Thread.sleep(100)
```
Comment (johanandren) [inline akka-typed-tests/src/test/scala/docs/akka/typed/GracefulStopDocSpec.scala]: please add a comment about why there are sleeps here, given that it is pulled into the docs. "Give the actor some time to spawn children" or something like that

File: akka-typed-tests/src/test/scala/docs/akka/typed/GracefulStopDocSpec.scala [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+    // brutally stop the system
+    system.terminate()
+
+    Thread.sleep(100)
```
Comment (johanandren) [inline akka-typed-tests/src/test/scala/docs/akka/typed/GracefulStopDocSpec.scala]: this one isn't needed

**Review by johanandren — COMMENTED** [review]
Looks good. Some minor things and then I'm missing some docs update where the samples are actually pulled into the docs?

File: akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (patriknw) [reply akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala]: I think we should use `.onSignal {`

It's similar as we prefer `collection.map {` over `collection map {`

File: akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala [hunk 10]
```diff
@@ -193,6 +192,33 @@ object Actor {
   def onSignal[T](handler: PartialFunction[(ActorContext[T], Signal), Behavior[T]]): Behavior[T] =
     immutable[T]((_, _) ⇒ same).onSignal(handler)
 
+  /**
+   * Behavior that stops all child actors, waits for them to stop and then
+   * runs some user defined cleanup before stopping itself.
+   */
+  def gracefulStop[T](cleanup: ActorContext[T] ⇒ Unit): Behavior[T] =
```
Comment (patriknw) [inline akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala]: what is the difference between `gracefulStop` and the existing stopped?

```
  /**
   * Return this behavior from message processing to signal that this actor
   * shall terminate voluntarily. If this actor has created child actors then
   * these will be stopped as part of the shutdown procedure.
   *
   * The PostStop signal that results from stopping this actor will be passed to the
   * given `postStop` behavior. All other messages and signals will effectively be
   * ignored.
   */
  def stopped[T](postStop: Behavior[T]): Behavior[T] = Behavior.stopped(postStop)
```

Isn't the `cleanup` callback like the `postStop` parameter?

Comment (gosubpl) [reply akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala]: `cleanup` is run before the actor is stopped. `postStop` runs after the actor has stopped. That is how I understand this. Please correct me if I am wrong.

File: akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala (hunk 6 above)
Comment (gosubpl) [reply akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala]: yes

File: akka-typed/src/main/scala/akka/typed/javadsl/Actor.scala (hunk 7 above)
Comment (gosubpl) [reply akka-typed/src/main/scala/akka/typed/javadsl/Actor.scala]: sure

File: akka-typed-tests/src/test/scala/docs/akka/typed/GracefulStopDocSpec.scala (hunk 9 above)
Comment (gosubpl) [reply akka-typed-tests/src/test/scala/docs/akka/typed/GracefulStopDocSpec.scala]: removed, thanks. Might need to put it back if we see some builds randomly failing.

File: akka-typed-tests/src/test/scala/docs/akka/typed/GracefulStopDocSpec.scala (hunk 8 above)
Comment (gosubpl) [reply akka-typed-tests/src/test/scala/docs/akka/typed/GracefulStopDocSpec.scala]: sure

File: akka-typed-tests/src/test/scala/docs/akka/typed/GracefulStopDocSpec.scala (hunk 9 above)
Comment (johanandren) [reply akka-typed-tests/src/test/scala/docs/akka/typed/GracefulStopDocSpec.scala]: Why would we need it, it's basically terminate() wait 100 mills, then wait 3 seconds more in await

**Comment by gosubpl** [comment]
@johanandren @patriknw great many thanks for your reviews!
Some general remarks:
- docs are not there yet - do you think just adding a paragraph in the `typed.md` would be ok?
- the `onSignal` is actually `Actor.onSignal` behavior factory added by @hseeberger in https://github.com/akka/akka/issues/23640 . I've attempted clarifying the situation by a comment and some small code refactor.

**Comment by gosubpl** [comment]
Included the commit comment pushed by @ktoso (thanks for fixing!) that got lost due to force-push.

**Comment by gosubpl** [comment]
According to the gitter discussion, I am removing the `gracefulStop` behavior factory from the Actor API in favour of doing it with `Actor.stopped(postStop)` . I am keeping two doc examples (that I will add to `typed.md`) and the spec for the "stopping actor gracefully" story.

**Review by johanandren — APPROVED** [review]

**Comment by patriknw** [comment]
I will rebase this.

**Comment by gosubpl** [comment]
thanks @patriknw

**Comment by patriknw** [comment]
Continued in https://github.com/akka/akka/pull/24346

**Pull request closed without merging**